### PR TITLE
Ultrafast spectroscopy: pump pulse generation

### PR DESCRIPTION
**Pull request by imaliyov**
Pump pulse HDF5 generation and plotting functions, based on the `DynaRun` objects.
Most of the functions are implemented in `/src/perturbopy/postproc/utils/spectra_... .py` files.
TODO: pump pulse postprocessing.

### CODE REVIEW

**Comment by imaliyov** [comment]
@hurricane642 and @ltan01 the work with spectroscopy is not finished. But the part of the pulse generation is finished.

Since I've touched some other part of the code (including the units bug and some plotting), I thought I could merge asap.

**Comment by hurricane642** [comment]
> @hurricane642 and @ltan01 the work with spectroscopy is not finished. But the part of the pulse generation is finished.
> 
> Since I've touched some other parts of the codebase (including the energy units conversion bug and improved some plotting functions), I thought I could merge asap.

Sure, make sense, I'll make the review asap

File: src/perturbopy/postproc/utils/spectra_plots.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+    Parameters
+    ----------
+
+    anim_time : float
+        Animation time.
+
+    ax : matplotlib.axes.Axes
+        Axis for plotting the pump pulse excitation.
+    """
```
Comment (ltan01) [inline src/perturbopy/postproc/utils/spectra_plots.py:116]: Hi Ivan, everything looks great! I think my only comment is adding more documentation about the input parameters for these functions.

File: src/perturbopy/postproc/utils/spectra_generate_pulse.py [hunk 2]
```diff
@@ -0,0 +1,373 @@
...
+
+    return sigma
+
+
+def delta_occs_pulse_coef(t, dt, tw, sigma):
+    """
+    Additional occupation due to the pulse excitation.
+    Assuming the Gaussian pulse shape, the occupation is increasing in
+    time according to the error function.
+    """
```
Comment (ltan01) [inline src/perturbopy/postproc/utils/spectra_generate_pulse.py:46]: Also, I think it might be nice to add some input parameter documentation for these functions. Although these might not be as urgent since they're called within the script.

File: docs/perturbo_docs/input_parameters.py
Comment (hurricane642) [inline docs/perturbo_docs/input_parameters.py:1]: 💯

File: setup.cfg [hunk 3]
```diff
@@ -1,3 +1,3 @@
 [flake8]
 max-line-length = 160
-ignore = E114, E127, E124, E221, W293, E721, W503, F841, F401
+ignore = E114, E127, E124, E221, W293, E721, W503, F841, F401, E202
```
Comment (hurricane642) [inline setup.cfg:3]: Why do you prefer to ignore this error?

File: src/perturbopy/postproc/utils/spectra_generate_pulse.py [hunk 4]
```diff
@@ -0,0 +1,373 @@
...
+    Gaussian sigma: f(t) = A / ( sigma * sqrt(2pi)) exp(-(t-t0)^2 / (2sigma^2))
+    """
+
+    sigma = 1.0 / (2.0 * np.sqrt(2.0 * np.log(2.0))) * fwhm
+
+    return sigma
+
+
+def delta_occs_pulse_coef(t, dt, tw, sigma):
+    """
```
Comment (hurricane642) [inline src/perturbopy/postproc/utils/spectra_generate_pulse.py:42]: In this function, as well as two previous one, we need documentation

File: src/perturbopy/postproc/utils/spectra_generate_pulse.py [hunk 5]
```diff
@@ -0,0 +1,373 @@
...
+        Time step in fs for the pump pulse generation.
+        Note: the pump_time_step MUST match the one used in the dynamics run in Perturbo!
+
+    hole : bool, optional
+        If True, the pump pulse is for holes. Default is False (electrons).
+
+    finite_width : bool, optional
+        If True, the pulse is finite in time. If False, the pulse is a step function
+        and occs_amplitude will be set as initial occupation.
+    """
```
Comment (hurricane642) [inline src/perturbopy/postproc/utils/spectra_generate_pulse.py:92]: Need to have line with the `Return` - what we expect to obtain from this function

File: src/perturbopy/postproc/utils/spectra_plots.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+    plt.show()
+
+
+def animate_pump_pulse(time_step,
+                       elec_delta_occs_array, elec_kpoint_array, elec_energy_array,
+                       hole_delta_occs_array, hole_kpoint_array, hole_energy_array,
+                       pump_energy):
+    """
+    Animate the pump pulse excitation for electrons and holes.
+    """
```
Comment (hurricane642) [inline src/perturbopy/postproc/utils/spectra_plots.py:62]: In this and previous function more explicit docs is required

**Review by hurricane642 — CHANGES_REQUESTED** [review]
Hey Ivan, looks great!
I've only mention some questions related to the documentation, that's all

File: setup.cfg (hunk 3 above)
Comment (imaliyov) [reply setup.cfg:3]: Good question! :) 

This is the reson:
```
src/perturbopy/test_utils/compare_data/h5.py:54:64: E202 whitespace before '}'
        diff_str = f'{diff:.1e}, {rdiff * 100:.1e}%, {scalar1 = }, {scalar2 = }'
```

flake8 (or PEP) does not want me to put a space before `}` which is totally fine in the plain code. However, for f-strings it does not make sense: this `f'{x = }'` is more stylistically correct than this ` f'{x=}'`. Because for example, in the plain code, this `x=5` would be a PEP violation and this `x = 5` would not.

What do you think?

File: src/perturbopy/postproc/utils/spectra_plots.py (hunk 1 above)
Comment (imaliyov) [reply src/perturbopy/postproc/utils/spectra_plots.py:116]: I did it, thank you Laurie.

We should however discuss and establish some policies for documentation of functions in Perturbopy. 
Cases we sould try to avoid is for one-line function having 20 lines of documentation. This could make the code actuallly less readable.

The main functions and modules must be very well documented. Other functions, less. I still documented entirely, to make the PR asap and to postpone this question after the code release :)
Also, I am very well aware that if we ease this rule now, some people can stop documenting everything, so we have to intoduce this rule with caution :) 

 @hurricane642 @alex2shiyu 
 
 Below is a screenshot of a function, where the comments, in my opinion, make it less readable
 
image

File: src/perturbopy/postproc/utils/spectra_generate_pulse.py (hunk 2 above)
Comment (imaliyov) [reply src/perturbopy/postproc/utils/spectra_generate_pulse.py:46]: Here is the contrary case, where documentation is very much needed! 
Thank you, done!

File: src/perturbopy/postproc/utils/spectra_generate_pulse.py (hunk 4 above)
Comment (imaliyov) [reply src/perturbopy/postproc/utils/spectra_generate_pulse.py:42]: Done

File: src/perturbopy/postproc/utils/spectra_generate_pulse.py (hunk 5 above)
Comment (imaliyov) [reply src/perturbopy/postproc/utils/spectra_generate_pulse.py:92]: Good point! Done!

File: src/perturbopy/postproc/utils/spectra_plots.py (hunk 6 above)
Comment (imaliyov) [reply src/perturbopy/postproc/utils/spectra_plots.py:62]: I've done this, but don't think this is needed, see the comment above.
Still done it to accelerate the PR, so we can discuss later for future developments.

File: src/perturbopy/postproc/utils/constants.py [hunk 7]
```diff
@@ -14,7 +14,7 @@
                       'J': ['j', 'joule', 'joules'],
                       'Ry': ['ry', 'rydberg', 'rydbergs']}
 
-energy_units_vals = {'eV': (2.7211396132, 1), 'Ha': (1, 0), 'J': (4.359748199, -18), 'Ry': (0.5, 0)}
+energy_units_vals = {'eV': (2.7211396132, 1), 'Ha': (1, 0), 'J': (4.359748199, -18), 'Ry': (2.0, 0)}
```
Comment (jyyao27) [inline src/perturbopy/postproc/utils/constants.py:17]: What is the reason of this change?

Comment (imaliyov) [reply src/perturbopy/postproc/utils/constants.py:17]: The conversion was wrong before.

energy_conversion_factor('Ry', 'eV')
after the fix it gives:
>>> ppy.utils.constants.energy_conversion_factor('Ry', 'eV')
13.605698066

Before the fix, there was a wrong number

**Review by jyyao27 — APPROVED** [review]
All looks good to me now.

**Comment by ltan01** [comment]
Thank you for making the changes, Ivan. Everything looks good to me!

**Review by ltan01 — APPROVED** [review]
Thank you for making the changes, Ivan. Everything looks good to me!

File: setup.cfg (hunk 3 above)
Comment (hurricane642) [reply setup.cfg:3]: Agree, totally make sense

File: src/perturbopy/postproc/utils/spectra_plots.py (hunk 1 above)
Comment (hurricane642) [reply src/perturbopy/postproc/utils/spectra_plots.py:187]: That's interesting question for discussion. Maybe be we can try to make a convention for documentation of this type of internal functions in the more condensed and short matter. We'll think about this later, but @imaliyov , thank you a lot for documenting all of this now, that is great!
